### DISCUSSION TITLE
Can I insert historical data into Mimir database?

### DISCUSSION THREAD

**Discussion by schoettl** [grafana/mimir#1616] [General]
I couldn't find out yet if Mimir allows deferred insertion of historical data.

In my use-case, I can't guarantee that metrics data is inserted immediately because there may be a broken internet connection. But I still need the data in the TSDB for later analysis.

In VictoriaMetrics this should be possible.

**Comment by pstibrany** [comment]
If you have historical data in TSDB blocks, it is possible to upload them into Mimir's storage. See https://grafana.com/docs/mimir/latest/migration-guide/migrating-from-thanos-or-prometheus/#uploading-historic-blocks-to-the-grafana-mimir-storage-bucket for details.

If your data is in different format, you will need to create TSDB blocks first, for example using Promtool backfilling process (https://prometheus.io/docs/prometheus/latest/storage/#backfilling-from-openmetrics-format), and then you can upload blocks into Mimir's storage.

**Reply by schoettl** [reply]
Thanks @pstibrany for the links! So I understand that Mimir can read the Prometheus TSDB block format and I only need to copy the files. And to create the TSDB blocks later, I can use Prometheus backfilling process.

BTW, did I get this right: Prometheus format and OpenMetrics format are pretty much compatible? Current versions of Prometheus exporter libraries probably generate valid OpenMetrics?

**Reply by pstibrany** [reply]
Yes, Mimir uses Prometheus TSDB format internally. Copying files to Mimir's storage (bucket) and fixing metadata as described in the first link, makes data visible to Mimir for querying.

Prometheus exposition format and OpenMetrics are very similar but there are small changes. However as you say, some client libraries use OpenMetrics format now.
